### DISCUSSION TITLE
Dynamic x-axis

### DISCUSSION THREAD

**Discussion by zell-mbc** [patrykandpatrick/vico#891] [Questions]
### Question

My charts (v2.0.0-beta.1) are based on a timeline which may span multiple years. I am collecting millisecond timestamps which are mapped to a precision of one minute along your proposal here: https://patrykandpatrick.com/vico/wiki/cartesian-charts/value-formatting/
That works.

Now, depending on the time span of the data at hand I decide if the horizontal labels are formatted as "yyyy MMM" or "MMM dd", this works as well.
But what I really want is for this to be dynamic, based on what's on the screen. One might start with showing a year of data where "MMM dd" fits well, and as you zoom in, maybe into one particular week, I'd like to change to e.g. "dd HH:mm", down to the maximum resolution of 1 minute and "HH:mm"

What I think I need for this would be the first and last item which is shown on screen. Is there any way to retrieve these values inside a CartesianValueFormatter? Or maybe there is a better way to achieve this?

### Vico version(s)

v2.0.0-beta.1

### UI framework(s)

Jetpack Compose

**Comment by shrift** [comment]
I am interested in similar behavior. In addition to wanting to change timestamp formatting, I want to be able to show in an external compose view what the current time range is for the currently visible data in the chart.
In my case we are preferring to not allow zoom, only allowing horizontal scrolling, and instead modify xStep values to affect the "density" of the data in the current view based on external buttons. The intricacies of the specifics of my implementation aside, is there a way to achieve what OP asked about? Receiving the first and last items shown on screen as the user zooms / scrolls?

**Comment by adrcotfas** [comment]
Hey @zell-mbc, @shrift
Did you find a solution for this? I'm also interested in formatting the x values based on the visible items.

**Comment by Shusek** [comment]
In my case, I would like to have the ability to calculate `minY` and `maxY` for the "RangeProvider" not for the entire collection but only for the portion visible to the user.

**Comment by adrcotfas** [comment]
Hey,
Here's an example of what I would like to achieve, a dynamic label for the first visible item of a horizontal axis.
Is this possible by extending existing components?

https://github.com/user-attachments/assets/9f9d4108-283c-4e62-9375-46cce611146b

**Reply by Gowsky** [reply]
Hello, @adrcoftas. Yes, you can achieve this by subclassing `HorizontalAxis`.

**Comment by Gowsky** [comment]
Hello, @zell-mbc. `VicoScrollState` and `VicoZoomState` don't currently expose the visible _x_-range, but we plan on adding this in the future. In the meantime, you can retrieve it from, for example, `HorizontalAxis.ItemPlacer`. Please see [here](https://github.com/patrykandpatrick/vico/discussions/788#discussioncomment-11182080).

Please note that formatting labels based on the zoom factor in a way that changes their length is valid only if the zoom factor doesn't, in turn, depend on the length of the labels. If these two things were true at once (which happens, for example, with `Zoom.Content` and `HorizontalAxis.ItemPlacer.aligned` with `addExtremeLabelPadding = true`), then a circular dependency would occur.
